scripts/licenses/licenses/library.go

```diff
@@ -0,0 +1,175 @@
+		return nil, err
+	}
+	pkgsByLicense := make(map[string][]*build.Package)
+	for _, d := range deps {
+		if isStdLib(d) {
+			// No license requirements for the Go standard library.
+			continue
+		}
+		licensePath, err := Find(d)
+		if err != nil {
+			glog.Errorf("Failed to find license for %s: %v", d.ImportPath, err)
+		}
+		if licensePath == pkgLicense {
+			// Skip dependencies that share a license file with pkg.
+			// These are not external libraries.
+			continue
+		}
+		pkgsByLicense[licensePath] = append(pkgsByLicense[licensePath], d)
+	}
+	var libraries []*Library
+	for licensePath, pkgs := range pkgsByLicense {
+		if licensePath == "" {
+			// No license for these packages - return each one as a separate library.
+			for _, pkg := range pkgs {
+				libraries = append(libraries, &Library{
+					Packages: []*build.Package{pkg},
+				})
+			}
+			continue
+		}
+		libraries = append(libraries, &Library{
+			LicensePath: licensePath,
+			Packages:    pkgs,
+		})
+	}
+	return libraries, nil
+}
+
+// Name is the common prefix of the import paths for all of the packages in this library.
+func (l *Library) Name() string {
+	if len(l.Packages) == 0 {
+		return ""
+	}
+	if len(l.Packages) == 1 {
+		return l.Packages[0].ImportPath
+	}
+	var importPaths []string
+	for _, pkg := range l.Packages {
+		importPaths = append(importPaths, pkg.ImportPath)
+	}
+	sort.Strings(importPaths)
+	min, max := importPaths[0], importPaths[len(importPaths)-1]
+	lastSlashIndex := 0
+	for i := 0; i < len(min) && i < len(max); i++ {
+		if min[i] != max[i] {
+			return min[:lastSlashIndex]
+		}
+		if min[i] == '/' {
+			lastSlashIndex = i
+		}
+	}
+	return min
+}
+
+func (l *Library) String() string {
+	return l.Name()
+}
+
+// ImportPackage returns information about the package identified by the given import path.
+// If there is a "vendor" directory in workingDir, packages in that directory will take precedence
+// over packages with the same import path found elsewhere.
+func importPackage(ctx *build.Context, importPath string, workingDir string) (*build.Package, error) {
+	cacheKey := workingDir + ":" + importPath
+	if pkg, ok := pkgCache.Load(cacheKey); ok {
+		return pkg.(*build.Package), nil
+	}
+
+	pkg, err := ctx.Import(importPath, workingDir, 0)
+	if err != nil {
+		return nil, err
+	}
+
+	pkgCache.Store(cacheKey, pkg)
+	return pkg, nil
+}
+
+// IsStdLib returns true if this package is part of the Go standard library.
+func isStdLib(pkg *build.Package) bool {
+	return pkg.Root == build.Default.GOROOT
+}
+
+// Dependencies finds the Go packages used by this package, directly or transitively.
+// They are added to the provided deps map.
+func dependencies(ctx *build.Context, pkg *build.Package, deps map[string]*build.Package) error {
+	for _, imp := range pkg.Imports {
+		if imp == "C" {
+			return fmt.Errorf("%s has a dependency on C code, which cannot be inspected for further dependencies", pkg.ImportPath)
+		}
+		if _, ok := deps[imp]; !ok {
```

Comment:
nit: outdent the main code flow (i.e. have `if ok { continue }`)

Reply:
Done.